src/initialize.cpp

```diff
@@ -0,0 +1,291 @@
+#include "openmc/initialize.h"
+
+#include <cstddef>
+#include <cstdlib> // for getenv
+#include <cstring>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+#include "openmc/capi.h"
+#include "openmc/constants.h"
+#include "openmc/cross_sections.h"
+#include "openmc/error.h"
+#include "openmc/geometry_aux.h"
+#include "openmc/hdf5_interface.h"
+#include "openmc/material.h"
+#include "openmc/message_passing.h"
+#include "openmc/mgxs_interface.h"
+#include "openmc/nuclide.h"
+#include "openmc/output.h"
+#include "openmc/plot.h"
+#include "openmc/random_lcg.h"
+#include "openmc/settings.h"
+#include "openmc/simulation.h"
+#include "openmc/string_utils.h"
+#include "openmc/summary.h"
+#include "openmc/tallies/tally.h"
+#include "openmc/thermal.h"
+#include "openmc/timer.h"
+
+
+int openmc_init(int argc, char* argv[], const void* intracomm)
+{
+  using namespace openmc;
+
+#ifdef OPENMC_MPI
+  // Check if intracomm was passed
+  MPI_Comm comm;
+  if (intracomm) {
+    comm = *static_cast<const MPI_Comm *>(intracomm);
+  } else {
+    comm = MPI_COMM_WORLD;
+  }
+
+  // Initialize MPI for C++
+  initialize_mpi(comm);
+#endif
+
+  // Parse command-line arguments
+  int err = parse_command_line(argc, argv);
+  if (err) return err;
+
+  // Start total and initialization timer
+  simulation::time_total.start();
+  simulation::time_initialize.start();
+
+#ifdef _OPENMP
+  // If OMP_SCHEDULE is not set, default to a static schedule
+  char* envvar = std::getenv("OMP_SCHEDULE");
+  if (!envvar) {
+    omp_set_schedule(omp_sched_static, 0);
+  }
+#endif
+
+  // Initialize random number generator -- if the user specifies a seed, it
+  // will be re-initialized later
+  openmc::openmc_set_seed(DEFAULT_SEED);
+
+  // Read XML input files
+  read_input_xml();
+
+  // Check for particle restart run
+  if (settings::particle_restart_run) settings::run_mode = RUN_MODE_PARTICLE;
+
+  // Stop initialization timer
+  simulation::time_initialize.stop();
+
+  return 0;
+}
+
+namespace openmc {
+
+#ifdef OPENMC_MPI
+void initialize_mpi(MPI_Comm intracomm)
+{
+  mpi::intracomm = intracomm;
+
+  // Initialize MPI
+  int flag;
+  MPI_Initialized(&flag);
+  if (!flag) MPI_Init(nullptr, nullptr);
+
+  // Determine number of processes and rank for each
+  MPI_Comm_size(intracomm, &mpi::n_procs);
+  MPI_Comm_rank(intracomm, &mpi::rank);
+  mpi::master = (mpi::rank == 0);
+
+  // Create bank datatype
+  Particle::Bank b;
+  MPI_Aint disp[6];
+  MPI_Get_address(&b.r, &disp[0]);
+  MPI_Get_address(&b.u, &disp[1]);
+  MPI_Get_address(&b.E, &disp[2]);
+  MPI_Get_address(&b.wgt, &disp[3]);
+  MPI_Get_address(&b.delayed_group, &disp[4]);
+  MPI_Get_address(&b.particle, &disp[5]);
+  for (int i = 5; i >= 0; --i) disp[i] -= disp[0];
```

Comment:
To be consistent with the other commits, consider placing the disp[i] -= disp[0]; on a seperate line from the for loop.